fix circular objects in json reporter. Closes #2433

jeversmann opened the pull request:
In #2433, @boneskull recommended using `jsonStringify()` from `./lib/utils`, but that function isn't exported by the utils module, and more importantly it isn't actually circular-reference safe.

Instead, I took the custom replacer from [this StackOverflow post](http://stackoverflow.com/questions/11616630/json-stringify-avoid-typeerror-converting-circular-structure-to-json) and added it to the reporter's invocation of `JSON.stringify`.

The test for this issue is mostly a copy of the 1st json reporter test. It has fewer asserts because there's no way to test for the change to the object that is made by the replacer, but the test _does_ fail without the additional code.

Munter commented:
Nice one. I'm not a big fan of the duplication though. Should this be put in its own reusable util function?

jeversmann commented:
There's no duplication here? This one reporter is the only spot I'm aware of with this problem.

ScottFreeCode commented:
IMHO if the version in our `util` component isn't actually circular-safe we should fix it (or make it a reference to a dependency that's safe), and if it isn't exported it should be, so other components such as reporters can just call on the util function.

boneskull commented:
hmm, yes, it seems [canonicalize()](https://github.com/mochajs/mocha/blob/36184aa78593772032988efa8845f3dc996b5c6f/lib/utils.js#L582-L582) is circular-safe, which is what I thought I was pointing you to.  oops.  but that function is for displaying diffs.

boneskull commented:
It would *seem* a public API to output JSON--while avoiding circular references--would be helpful to reporter implementations.  That's tangential to #2433.

Looking more closely at the JSON reporter, it already [purports](https://github.com/mochajs/mocha/blob/2bb2b9fa35818db7a02e5068364b0c417436b1af/lib/reporters/json.js#L61-L61) to kill circular references.  The bug is that *it doesn't actually succeed in doing this*.

The JSON stream reporter works because it doesn't attempt to "stringify" an `Error`--it just yanks the `message` and `stack` properties out.

I think what this PR should do is actually very fine-tuned to this bug.  @jeversmann Can you instead modify [errorJSON()](https://github.com/mochajs/mocha/blob/2bb2b9fa35818db7a02e5068364b0c417436b1af/lib/reporters/json.js#L86-L86) or [clean()](https://github.com/mochajs/mocha/blob/2bb2b9fa35818db7a02e5068364b0c417436b1af/lib/reporters/json.js#L69) to skip objects it's already seen?  

I don't know what the "best" implementation would look like here, but we should retain as much of the `Error` as is feasible.  We shouldn't take the easy way out and do what the JSON stream reporter does. That would would be a breaking change, and not a bug fix.

jeversmann commented:
I moved the circular object logic into a helper function that's called in `clean` so it behaves according to the comment above it.

The 'gotcha' here is that `Error` objects don't `JSON.stringify` correctly, so formatting `test.err` is a two-step process which might behave strangely if a test throws an object that isn't an `instanceof Error` but contains one or more `Error`s as children. [The suggested fix seems to be defining `Error.prototype.toJSON`](http://stackoverflow.com/a/18391400) but that seems like overkill unless the current method causes problems for anyone.

ScottFreeCode commented:
> ...might behave strangely if a test throws an object that isn't an  instanceof Error  but contains one or more  Error s as children.

I thought Mocha doesn't accept non-`Error` failures in the first place?

jeversmann commented:
The problem in #2433 deals specifically with an object whose prototype chain contains `Error`, but the test I wrote to provoke the circular parsing problem just `throw`s a regular object with a circular reference in it.

I can modify the test to build the thrown object from an `Error` but the code to handle printing it shouldn't change.

ScottFreeCode commented:
Yeah, I'm not too concerned with having a test that focuses on the circular reference thing without worrying about rigging up a "real" `Error`; I just wasn't sure how the comment about handling the case of `Error` as a member of non-`Error` objects was relevant. Sorry for any confusion!

jeversmann commented:
@boneskull This is still labeled as pr-needs-work, but I think I've addressed the feedback.
The bot is right that this PR has become stale, but #2433 is still around, so I'd rather merge this in.

boneskull commented:
@jeversmann thanks for the heads-up.  someone will take a look at this, maybe even me!

xyleen commented:
Hey! I've stumbled upon this situation myself and the proposed commits seem to resolve the issue for me. 

Do you plan on merging this pull request sometime? Is this PR missing anything that prevents it from being mergable?

boneskull commented:
continued in #3318.  thanks

boneskull commented:
landed in https://github.com/mochajs/mocha/commit/741b0bd4eff23b2388b605c8f4cda99d2d7ea38a
